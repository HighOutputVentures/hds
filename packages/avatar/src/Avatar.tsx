export type AvatarProps = {};

export default function Avatar({}: AvatarProps) {
  return null;
}
