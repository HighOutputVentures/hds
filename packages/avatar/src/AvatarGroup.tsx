import {
  AvatarGroup,
  chakra,
  HStack,
  Icon,
  SystemStyleObject,
  Tooltip,
  useBreakpoint,
  useMultiStyleConfig,
} from '@chakra-ui/react';
import * as React from 'react';
import { useActualSize } from './hooks';
import PlusIcon from './icons/PlusIcon';
import { Breakpoints, ResponsiveSize } from './types';
import { findClosestBreakpoint } from './utils';

type AvatarGroupSize = 'xs' | 'sm' | 'md';

type AvatarGroupBaseProps = {
  max?: number | Partial<Record<Breakpoints, number>>;
  size?: AvatarGroupSize | ResponsiveSize<AvatarGroupSize>;
  hasAddButton?: boolean;
  onAddButtonClick?(): void;
};

export type AvatarGroupProps = SystemStyleObject & AvatarGroupBaseProps;

export default function HdsAvatarGroup(props: React.PropsWithChildren<AvatarGroupProps>) {
  const { max = 5, size = 'md', children, hasAddButton, onAddButtonClick, ...others } = props;

  const actualMax = useActualMax(max);
  const actualSize = useActualSize(size);

  return (
    <HStack sx={others} spacing="8px" /* retain spacing */>
      <AvatarGroup
        variant="hds"
        max={actualMax}
        size={size}
        spacing={{ xs: '-4px', sm: '-8px', md: '-12px' }[actualSize]}
        aria-label="Group of users"
      >
        {React.Children.map(children, (child, zIndex) => {
          if (!React.isValidElement(child)) return null;

          return React.cloneElement<any>(child, {
            size,
            zIndex,
            __bordered: true,
          });
        })}
      </AvatarGroup>

      {hasAddButton && <AddButton size={actualSize} onClick={onAddButtonClick} />}
    </HStack>
  );
}

type AddButtonProps = React.ComponentProps<'button'> & { size: string /* unsafe. ü§ê */ };

function AddButton({ size, ...props }: AddButtonProps) {
  const styles = useMultiStyleConfig('AvatarGroupButton', { size });

  return (
    <Tooltip
      variant="hds-avatar"
      hasArrow
      label="Add user"
      placement="top"
      role="tooltip"
      aria-label="Add user"
    >
      <chakra.button role="button" aria-label="Add User" sx={styles.container} {...props}>
        <Icon as={PlusIcon} sx={styles.icon} />
      </chakra.button>
    </Tooltip>
  );
}

function useActualMax(max: AvatarGroupProps['max'], fallback = 'md') {
  const breakpoint = useBreakpoint({ fallback });

  const breakpoints = typeof max === 'number' ? [] : Object.keys(max);

  return typeof max === 'number'
    ? max
    : breakpoint in max
    ? max[breakpoint]
    : breakpoints.length
    ? max[findClosestBreakpoint(breakpoints, breakpoint)]
    : fallback;
}
