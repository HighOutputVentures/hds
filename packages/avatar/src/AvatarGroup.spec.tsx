import '@testing-library/jest-dom';
import { fireEvent, render, waitFor } from '@testing-library/react';
import * as React from 'react';
import Avatar from './Avatar';
import AvatarGroup, { AvatarGroupProps } from './AvatarGroup';

jest.mock('@chakra-ui/react', function mockUseBreakpointHook() {
  return {
    ...jest.requireActual('@chakra-ui/react'),
    useBreakpoint() {
      return 'lg';
    },
  };
});

describe('AvatarGroup', () => {
  const renderTestComponent = (options: AvatarGroupProps = {}) => {
    return render(
      <AvatarGroup {...options}>
        <Avatar src="https://i.pravatar.cc/50?u=1" />
        <Avatar src="https://i.pravatar.cc/50?u=2" />
        <Avatar src="https://i.pravatar.cc/50?u=3" />
        <Avatar src="https://i.pravatar.cc/50?u=4" />
        <Avatar src="https://i.pravatar.cc/50?u=5" />
      </AvatarGroup>,
    );
  };

  it('Should render container', () => {
    const { queryByRole } = renderTestComponent({ max: Infinity });
    expect(queryByRole('group', { name: 'Group of users' })).toBeDefined();
  });

  it('Should render avatars', () => {
    const { queryAllByRole } = renderTestComponent({ max: Infinity });
    expect(queryAllByRole('img', { name: 'avatar' })).toHaveLength(5);
  });

  it('Should be able to set max avatars to show', () => {
    const max = 3;
    const { queryAllByRole } = renderTestComponent({ max });
    expect(queryAllByRole('img', { name: 'avatar' })).toHaveLength(max);
  });

  it('Should show excess label', () => {
    const max = 3;
    const { queryByText } = renderTestComponent({ max });
    expect(queryByText('+2')).toBeDefined();
  });

  it('Should be able to toggle add button', () => {
    const { queryByRole } = renderTestComponent({ hasAddButton: true });
    expect(queryByRole('button', { name: 'Add User' })).toBeInTheDocument();
  });

  it('Add button should call callback fn', async () => {
    const onAddButtonClick = jest.fn();
    const { getByRole } = renderTestComponent({ hasAddButton: true, onAddButtonClick });

    fireEvent.click(getByRole('button', { name: 'Add User' }));

    await waitFor(() => {
      expect(onAddButtonClick).toHaveBeenCalled();
    });
  });

  it('Should show tooltip when add button is hovered', async () => {
    const { getByRole, queryByRole } = renderTestComponent({ hasAddButton: true });

    fireEvent.mouseOver(getByRole('button', { name: 'Add User' }));

    await waitFor(() => {
      expect(queryByRole('tooltip', { name: 'Add user' })).toBeDefined();
    });
  });
});
