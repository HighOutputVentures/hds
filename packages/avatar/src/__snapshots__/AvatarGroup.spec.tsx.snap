// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup Snapshot Should match snapshot 1`] = `
<div
  className="chakra-stack css-16awuyr"
>
  <div
    aria-label="Group of users"
    className="chakra-avatar__group css-1rs5v3u"
    role="group"
  >
    <span
      className="chakra-avatar css-u2n1c6"
      data-bordered={true}
      data-clickable={false}
      data-elevated={false}
      data-fallback={false}
      data-online={false}
    >
      <svg
        aria-label=" avatar"
        className="chakra-icon css-rv5rod"
        fill="none"
        focusable={false}
        height="32"
        role="img"
        viewBox="0 0 32 32"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.6666 28C26.6666 26.1392 26.6666 25.2089 26.4369 24.4518C25.9199 22.7473 24.586 21.4134 22.8814 20.8963C22.1244 20.6667 21.194 20.6667 19.3332 20.6667H12.6666C10.8058 20.6667 9.87546 20.6667 9.1184 20.8963C7.41386 21.4134 6.07997 22.7473 5.5629 24.4518C5.33325 25.2089 5.33325 26.1392 5.33325 28M21.9999 10C21.9999 13.3137 19.3136 16 15.9999 16C12.6862 16 9.99992 13.3137 9.99992 10C9.99992 6.68629 12.6862 4 15.9999 4C19.3136 4 21.9999 6.68629 21.9999 10Z"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2.66667"
        />
      </svg>
    </span>
    <span
      className="chakra-avatar css-1g0pkd1"
      data-bordered={true}
      data-clickable={false}
      data-elevated={false}
      data-fallback={false}
      data-online={false}
    >
      <svg
        aria-label=" avatar"
        className="chakra-icon css-rv5rod"
        fill="none"
        focusable={false}
        height="32"
        role="img"
        viewBox="0 0 32 32"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.6666 28C26.6666 26.1392 26.6666 25.2089 26.4369 24.4518C25.9199 22.7473 24.586 21.4134 22.8814 20.8963C22.1244 20.6667 21.194 20.6667 19.3332 20.6667H12.6666C10.8058 20.6667 9.87546 20.6667 9.1184 20.8963C7.41386 21.4134 6.07997 22.7473 5.5629 24.4518C5.33325 25.2089 5.33325 26.1392 5.33325 28M21.9999 10C21.9999 13.3137 19.3136 16 15.9999 16C12.6862 16 9.99992 13.3137 9.99992 10C9.99992 6.68629 12.6862 4 15.9999 4C19.3136 4 21.9999 6.68629 21.9999 10Z"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2.66667"
        />
      </svg>
    </span>
    <span
      className="chakra-avatar css-11f37eh"
      data-bordered={true}
      data-clickable={false}
      data-elevated={false}
      data-fallback={false}
      data-online={false}
    >
      <svg
        aria-label=" avatar"
        className="chakra-icon css-rv5rod"
        fill="none"
        focusable={false}
        height="32"
        role="img"
        viewBox="0 0 32 32"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.6666 28C26.6666 26.1392 26.6666 25.2089 26.4369 24.4518C25.9199 22.7473 24.586 21.4134 22.8814 20.8963C22.1244 20.6667 21.194 20.6667 19.3332 20.6667H12.6666C10.8058 20.6667 9.87546 20.6667 9.1184 20.8963C7.41386 21.4134 6.07997 22.7473 5.5629 24.4518C5.33325 25.2089 5.33325 26.1392 5.33325 28M21.9999 10C21.9999 13.3137 19.3136 16 15.9999 16C12.6862 16 9.99992 13.3137 9.99992 10C9.99992 6.68629 12.6862 4 15.9999 4C19.3136 4 21.9999 6.68629 21.9999 10Z"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2.66667"
        />
      </svg>
    </span>
    <span
      className="chakra-avatar css-13lgae6"
      data-bordered={true}
      data-clickable={false}
      data-elevated={false}
      data-fallback={false}
      data-online={false}
    >
      <svg
        aria-label=" avatar"
        className="chakra-icon css-rv5rod"
        fill="none"
        focusable={false}
        height="32"
        role="img"
        viewBox="0 0 32 32"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.6666 28C26.6666 26.1392 26.6666 25.2089 26.4369 24.4518C25.9199 22.7473 24.586 21.4134 22.8814 20.8963C22.1244 20.6667 21.194 20.6667 19.3332 20.6667H12.6666C10.8058 20.6667 9.87546 20.6667 9.1184 20.8963C7.41386 21.4134 6.07997 22.7473 5.5629 24.4518C5.33325 25.2089 5.33325 26.1392 5.33325 28M21.9999 10C21.9999 13.3137 19.3136 16 15.9999 16C12.6862 16 9.99992 13.3137 9.99992 10C9.99992 6.68629 12.6862 4 15.9999 4C19.3136 4 21.9999 6.68629 21.9999 10Z"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2.66667"
        />
      </svg>
    </span>
    <span
      className="chakra-avatar css-1cflss4"
      data-bordered={true}
      data-clickable={false}
      data-elevated={false}
      data-fallback={false}
      data-online={false}
    >
      <svg
        aria-label=" avatar"
        className="chakra-icon css-rv5rod"
        fill="none"
        focusable={false}
        height="32"
        role="img"
        viewBox="0 0 32 32"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M26.6666 28C26.6666 26.1392 26.6666 25.2089 26.4369 24.4518C25.9199 22.7473 24.586 21.4134 22.8814 20.8963C22.1244 20.6667 21.194 20.6667 19.3332 20.6667H12.6666C10.8058 20.6667 9.87546 20.6667 9.1184 20.8963C7.41386 21.4134 6.07997 22.7473 5.5629 24.4518C5.33325 25.2089 5.33325 26.1392 5.33325 28M21.9999 10C21.9999 13.3137 19.3136 16 15.9999 16C12.6862 16 9.99992 13.3137 9.99992 10C9.99992 6.68629 12.6862 4 15.9999 4C19.3136 4 21.9999 6.68629 21.9999 10Z"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2.66667"
        />
      </svg>
    </span>
  </div>
</div>
`;
